File: SW_Expert_Academy/구미3반/Flatten.cpp
#include <iostream>
#include <cstring>
using namespace std;
using ll = long long;
int main(void) {
    //freopen("input.txt", "r", stdin);
    int n,k,dump,tmp,max_val,min_val; 
    int cnt[101];
    for (int t = 1; t <= 10; ++t) {
        memset(cnt,0,sizeof(cnt)); max_val=-1;min_val=1001;
        cin >> dump;
        for (int i = 0 ; i < 100 ; ++i) {
            cin >> tmp; 
            ++cnt[tmp];
            min_val = min(min_val,tmp);
            max_val = max(max_val,tmp);
        }
        while( min_val < max_val)
        {   
            int diff = min(dump,min(cnt[max_val],cnt[min_val]));
            cnt[max_val-1]+=diff;
            cnt[min_val+1]+=diff;
            cnt[max_val]-=diff;
            cnt[min_val]-=diff;
            dump -= diff;
            if (cnt[max_val] == 0) --max_val;
            if (cnt[min_val] == 0) ++min_val;

            if (dump==0) break;
        }
        
        cout <<'#' << t <<' ' <<(max_val - min_val)<<'\n';
    }

    return 0;
}

File: SW_Expert_Academy/구미3반/Flatten.py
# import sys
# sys.stdin = open('input.txt','r')
#from pprint import pprint as print
T = 10

for t in range(1, T + 1):
    dump = int(input())
    cnt = [0]*101
    lst = list(map(int,input().split()))
    max_val, min_val = max(lst), min(lst)
    for val in lst:
        cnt[val] += 1
        
    while min_val < max_val:
        if cnt[max_val] == 0:
            max_val-=1
            continue
        if cnt[min_val] == 0:
            min_val+=1
            continue
        
        diff = min(cnt[max_val],cnt[min_val],dump)
        cnt[max_val-1]+=diff
        cnt[max_val]-=diff
        cnt[min_val]-=diff
        cnt[min_val+1]+=diff
        dump-=diff
        
        if not cnt[max_val]: max_val-=1
        if not cnt[min_val]: min_val+=1
        if not dump: break
    print(f'#{t} {max_val-min_val}')
    
        